### PR TITLE
feat: `have` vs `let` linter

### PR DESCRIPTION
**Pull request by adomani**
##  #12190 seems to have become the main PR for this linter.

The `have` vs `let` linter flags uses of `have` to introduce a hypothesis whose Type is not `Prop`

Here is a summary of the PR.  Besides changes `have --> let` this is everything else.

* `Mathlib/Analysis/Convolution.lean`: two parallel simplified proofs.
* `Mathlib/CategoryTheory/Limits/Final.lean`: a `have` that could not be replaced by `let`.
* `Mathlib/Combinatorics/SimpleGraph/Connectivity.lean`: vanishing `haveSuffices` due to `have --> let`.
* `Mathlib/LinearAlgebra/TensorProduct/Graded/Internal.lean`: `have := ...; exact this` becoming `exact ...`.
* `Mathlib/FieldTheory/Finite/GaloisField.lean`: using `let` makes the `unusedVariable` linter complain.
* `Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean`: unset linter since `clear_value` complains.
* `Mathlib/Tactic/NormNum/Inv.lean`: unset linter since `generalize` complains.

What is the syntax in `Mathlib/CategoryTheory/Extensive.lean`?

`bench` at #12180, testing only the replacements, not the linter itself.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by YaelDillies** [comment]
I'm slightly concerned that the `have`s you're turning into `let`s were declaring data that we didn't care about and that Lean now has the option of unfolding when we know it is unnecessary.
1. Could you benchmark this PR once CI is green?
2. Could the linter only run on definitions and incomplete proofs? If a proof is complete, we know for sure the `have`s can stay `have`s. In a definition however, a type-valued `have` is most certainly a mistake.

**Comment by adomani** [comment]
Yaël, with respect to benchmarking, I was definitely planning on doing that, once I finish up the missing replacements.

Running the linter only on incomplete proofs: this is something that can probably be done.  My first approach would probably be to run the linter only if the command emits a *warning*.  I would probably still not run the linter with *errors* since then the type information is probably unreliable.

File: Mathlib/Tactic/HaveLetLinter.lean [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        let _ : Ord MetavarDecl := { compare := (compare ·.index ·.index) }
+        -- we extract the `MetavarDecl` with largest index after a `have`, since this one
+        -- holds information about the metavariable where `have` introduces the new hypothesis.
+        let largestIdx := mvdecls.toArray.qsort (·.index > ·.index)
+        -- the relevant `LocalContext`
+        let lc := (largestIdx.getD 0 default).lctx
+        -- and the last declaration introduced: the one that `have` created
+        let ld := (lc.lastDecl.getD default).type
+        -- now, we get the `MetaM` state up and running to find the type of `ld`
+        let res ← liftCoreM do MetaM.run (ctx := { lctx := lc }) (s := { mctx := mctx }) <| do
```
Comment (eric-wieser) [inline Mathlib/Tactic/HaveLetLinter.lean:73]: Are you looking for [Lean.Elab.ContextInfo.runMetaM](https://leanprover-community.github.io/mathlib4_docs/Lean/Elab/InfoTree/Main.html#Lean.Elab.ContextInfo.runMetaM) here?

**Comment by eric-wieser** [comment]
Is it possible to split this PR into:
* Add the new linter, disabled-by-default
* Enable the linter and fix all the occurences

Scrolling through 150 files to find the new meta code is pretty annoying, especially over a train internet connection...

**Comment by adomani** [comment]
@eric-wieser, I opened #12190 that contains only the linter and the tests.  I did not disable the linter, but I also did not import the file anywhere.

**Comment by YaelDillies** [comment]
Actually, I don't really like my previous comment anymore. What about this instead?

The linter should send a warning on:
* a Prop-valued `let`
* a Type-valued `have` in an erroring branch of a proof

Note that the "Type-valued `have` in a completed sorry-free `def`" that I was referring to before is actually not bad since it's defeq to its value once the same is completed.

File: Mathlib/Tactic/HaveLetLinter.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Tactic/HaveLetLinter.lean:73]: Discussed further in https://github.com/leanprover-community/mathlib4/pull/12190

**Pull request closed without merging**